File: fly-http/src/main/java/com/xia/fly/http/body/upload/UIProgressResponseCallBack.java
package com.xia.fly.http.body.upload;

import android.os.Handler;
import android.os.Looper;
import android.os.Message;

import java.lang.ref.WeakReference;

/**
 * @author xia
 * @date 2018/8/3.
 * <p>
 * 描述：可以直接更新UI的回调
 */
@SuppressWarnings("WeakerAccess")
public abstract class UIProgressResponseCallBack implements ProgressResponseCallBack {
    private static final int RESPONSE_UPDATE = 0x02;

    //处理UI层的Handler子类
    private static class UIHandler extends Handler {
        //弱引用
        private final WeakReference<UIProgressResponseCallBack> mUIProgressResponseListenerWeakReference;

        public UIHandler(Looper looper, UIProgressResponseCallBack uiProgressResponseListener) {
            super(looper);
            mUIProgressResponseListenerWeakReference = new WeakReference<>(uiProgressResponseListener);
        }

        @Override
        public void handleMessage(Message msg) {
            switch (msg.what) {
                case RESPONSE_UPDATE:
                    final UIProgressResponseCallBack uiProgressResponseListener = mUIProgressResponseListenerWeakReference.get();
                    if (uiProgressResponseListener != null) {
                        //获得进度实体类
                        final ProgressModel progressModel = (ProgressModel) msg.obj;
                        //回调抽象方法
                        uiProgressResponseListener.onUIResponseProgress(progressModel.getCurrentBytes(),
                                progressModel.getContentLength(), progressModel.isDone());
                    }
                    break;
                default:
                    super.handleMessage(msg);
                    break;
            }
        }
    }

    //主线程Handler
    private final Handler mHandler = new UIHandler(Looper.getMainLooper(), this);

    @Override
    public void onResponseProgress(long bytesWritten, long contentLength, boolean done) {
        //通过Handler发送进度消息
        final Message message = Message.obtain();
        message.obj = new ProgressModel(bytesWritten, contentLength, done);
        message.what = RESPONSE_UPDATE;
        mHandler.sendMessage(message);
    }

    /**
     * UI层回调抽象方法
     *
     * @param bytesRead     当前读取响应体字节长度
     * @param contentLength 总字节长度
     * @param done          是否读取完成
     */
    public abstract void onUIResponseProgress(long bytesRead, long contentLength, boolean done);
}

File: fly-http/src/main/java/com/xia/fly/http/body/upload/UIProgressResponseCallBack.kt
package com.xia.fly.http.body.upload

import android.os.Handler
import android.os.Looper
import android.os.Message

import java.lang.ref.WeakReference

/**
 * @author xia
 * @date 2018/8/3.
 *
 *
 * 描述：可以直接更新UI的回调
 */
@Suppress("unused")
abstract class UIProgressResponseCallBack : ProgressResponseCallBack {

    //主线程Handler
    private val mHandler = UIHandler(Looper.getMainLooper(), apply {})

    //处理UI层的Handler子类
    private class UIHandler(looper: Looper, uiProgressResponseListener: UIProgressResponseCallBack) : Handler(looper) {
        //弱引用
        private val weakReference: WeakReference<UIProgressResponseCallBack> = WeakReference(uiProgressResponseListener)

        override fun handleMessage(msg: Message) {
            when (msg.what) {
                RESPONSE_UPDATE -> {
                    val uiProgressResponseListener = weakReference.get()
                    if (uiProgressResponseListener != null) {
                        //获得进度实体类
                        val progressModel = msg.obj as ProgressModel
                        //回调抽象方法
                        uiProgressResponseListener.onUIResponseProgress(progressModel.getCurrentBytes(),
                                progressModel.getContentLength(), progressModel.isDone())
                    }
                }
                else -> super.handleMessage(msg)
            }
        }
    }

    override fun onResponseProgress(bytesWritten: Long, contentLength: Long, done: Boolean) {
        //通过Handler发送进度消息
        val message = Message.obtain()
        message.obj = ProgressModel(bytesWritten, contentLength, done)
        message.what = RESPONSE_UPDATE
        mHandler.sendMessage(message)
    }

    /**
     * UI层回调抽象方法
     *
     * @param bytesRead     当前读取响应体字节长度
     * @param contentLength 总字节长度
     * @param done          是否读取完成
     */
    abstract fun onUIResponseProgress(bytesRead: Long, contentLength: Long, done: Boolean)

    companion object {
        private const val RESPONSE_UPDATE = 0x02
    }
}
